engine/src/main/java/io/camunda/zeebe/engine/state/message/TransientPendingSubscriptionState.java

```diff
@@ -0,0 +1,54 @@
+/*
+ * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH under
+ * one or more contributor license agreements. See the NOTICE file distributed
+ * with this work for additional information regarding copyright ownership.
+ * Licensed under the Zeebe Community License 1.1. You may not use this file
+ * except in compliance with the Zeebe Community License 1.1.
+ */
+package io.camunda.zeebe.engine.state.message;
+
+import java.util.Map;
+import java.util.Map.Entry;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.stream.Collectors;
+
+/**
+ * This class is used by {@link PendingProcessMessageSubscriptionState} and {@link
+ * PendingMessageSubscriptionState} to keep track of pending (process) message subscriptions. {@link
+ * PendingSubscription PendingSubscriptions} are added with a last sent time. The time can be
+ * updated with {@link #update(PendingSubscription, long)}. Pending subscriptions are retrieved with
+ * {@link #entriesBefore(long)}, ordered by the last sent time. <br>
+ * This class is thread safe. It's intended use is that one thread is adding, removing and updating
+ * entries while another thread is "observing" the entries by calling {@link #entriesBefore(long)}.
+ * <br>
+ * The iterable returned by {@link #entriesBefore(long)} may return outdated entries (i.e. entries
+ * which last sent time was updated so that those entries should not have been returned) and already
+ * removed entries. For the intended use, this is not a problem.
+ */
+public final class TransientPendingSubscriptionState {
+
+  // Reconsider thread-safety implications when changing the map implementation.
+  private final Map<PendingSubscription, Long> pending = new ConcurrentHashMap<>();
+
+  public void add(final PendingSubscription pendingSubscription, final long lastSentTime) {
+    pending.put(pendingSubscription, lastSentTime);
+  }
+
+  public void update(final PendingSubscription pendingSubscription, final long lastSentTime) {
+    pending.put(pendingSubscription, lastSentTime);
+  }
+
+  public void remove(final PendingSubscription pendingSubscription) {
+    pending.remove(pendingSubscription);
+  }
+
+  Iterable<PendingSubscription> entriesBefore(final long deadline) {
+    return pending.entrySet().stream()
+        .sorted(Map.Entry.comparingByValue())
+        .takeWhile(entry -> entry.getValue() < deadline)
+        .map(Entry::getKey)
+        .collect(Collectors.toList());
```

Comment:
❌ I don't think we can safely stream over this set while modifying the concurrent hashmap in another thread.

>Returns a Set view of the mappings contained in this map. **The set is backed by the map, so changes to the map are reflected in the set, and vice-versa**. The set supports element removal, which removes the corresponding mapping from the map, via the Iterator.remove, Set.remove, removeAll, retainAll, and clear operations.
The view's iterators and spliterators are weakly consistent.
The view's spliterator reports Spliterator.CONCURRENT, Spliterator.DISTINCT, and Spliterator.NONNULL.

We may need to make a copy of the entrySet before iterating, to avoid (from `java.util.Map`):
> If the map is modified while an iteration over the set is in progress (except through the iterator's own remove operation, or through the setValue operation on a map entry returned by the iterator) **the results of the iteration are undefined**.

Reply:
It's definitely confusing and the docs are a bit dense but I'm pretty confident that this is acceptable use.

Our usage pattern is the following: One thread is adding and removing entries. Another thread is iterating over (parts of) the entry set.

To quote again from the `ConcurrentHashMap` documentation:
> Similarly, Iterators, Spliterators and Enumerations return elements reflecting the state of the hash table at some point at or since the creation of the iterator/enumeration. They do not throw ConcurrentModificationException. However, iterators are designed to be used by only one thread at a time

In particular the `entrySet()` documentation says:
> The view's iterators and spliterators are weakly consistent.

That means:

> Most concurrent Collection implementations (including most Queues) also differ from the usual java.util conventions in that their Iterators and Spliterators provide weakly consistent rather than fast-fail traversal:
> - they may proceed concurrently with other operations
> - they will never throw ConcurrentModificationException
> - they are guaranteed to traverse elements as they existed upon construction exactly once, and may (but are not guaranteed to) reflect any modifications subsequent to construction.

So at worst, the reader thread may read (or not read) elements that were added (or removed) by the writer thread _after_ the reader thread has started iteration of the entry set.

Unless I'm missing something, this is acceptable for our use case.